aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs

```diff
@@ -89,21 +97,97 @@ pub fn pop_value_by_type(ty_arg: &Type, args: &mut VecDeque<Value>) -> SafeNativ
     }
 }
 
+// To aviod checking is_string_type multiple times, check type_arg only once, and convert into this enum
+enum SnapshotType {
+    U128,
+    U64,
+    String,
+}
+
+impl SnapshotType {
+    fn from_ty_args(context: &SafeNativeContext, ty_arg: &Type) -> SafeNativeResult<Self> {
+        match ty_arg {
+            Type::U128 => Ok(Self::U128),
+            Type::U64 => Ok(Self::U64),
+            _ => {
+                // Check if the type is a string
+                if is_string_type(context, ty_arg)? {
+                    Ok(Self::String)
+                } else {
+                    // If not a string, return an error
+                    Err(SafeNativeError::Abort {
+                        abort_code: EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE,
+                    })
+                }
+            },
+        }
+    }
+
+    pub fn pop_snapshot_field_by_type(
+        &self,
+        args: &mut VecDeque<Value>,
+    ) -> SafeNativeResult<SnapshotValue> {
+        self.parse_snapshot_value_by_type(aggregator_snapshot_field_value(&safely_pop_arg!(
+            args, StructRef
+        ))?)
+    }
+
+    pub fn pop_snapshot_value_by_type(
+        &self,
+        args: &mut VecDeque<Value>,
+    ) -> SafeNativeResult<SnapshotValue> {
+        match self {
+            SnapshotType::U128 => Ok(SnapshotValue::Integer(safely_pop_arg!(args, u128))),
+            SnapshotType::U64 => Ok(SnapshotValue::Integer(safely_pop_arg!(args, u64) as u128)),
+            SnapshotType::String => {
+                let input = string_to_bytes(safely_pop_arg!(args, Struct))?;
+                Ok(SnapshotValue::String(input))
+            },
+        }
+    }
+
+    pub fn parse_snapshot_value_by_type(&self, value: Value) -> SafeNativeResult<SnapshotValue> {
+        // Simpler to wrap to be able to reuse safely_pop_arg functions
+        let mut dummy_args = VecDeque::new();
+        dummy_args.push_back(value);
+        self.pop_snapshot_value_by_type(&mut dummy_args)
+    }
+
+    pub fn create_snapshot_value_by_type(&self, value: &SnapshotValue) -> SafeNativeResult<Value> {
+        match (self, value) {
+            (SnapshotType::U128, SnapshotValue::Integer(v)) => Ok(Value::u128(*v)),
+            (SnapshotType::U64, SnapshotValue::Integer(v)) => Ok(Value::u64(u128_to_u64(*v)?)),
+            (SnapshotType::String, _) => Ok(Value::struct_(Struct::pack(vec![Value::vector_u8(
+                match value {
+                    SnapshotValue::String(v) => v.clone(),
+                    SnapshotValue::Integer(v) => v.to_string().into_bytes(),
+                },
+            )]))),
+            // ty_arg cannot be Integer, if value is String
+            _ => Err(SafeNativeError::Abort {
+                abort_code: EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE,
+            }),
+        }
+    }
+}
+
+fn get_context_data<'t, 'b>(
+    context: &'t mut SafeNativeContext<'_, 'b, '_, '_>,
+) -> (&'b dyn AggregatorResolver, RefMut<'t, AggregatorData>) {
```

Comment:
Are we saving any lines of code with this function call? We are basically replacing a 1-line function call `context.extensions().get::<NativeAggregatorContext>();` with a 1-line function call `get_context_data()`;

Reply:
I found it cleaner to do :

```
let (_, mut aggregator_data) = get_context_data(context);
```
than 
```
       let aggregator_context = context.extensions().get::<NativeAggregatorContext>();
        let mut aggregator_data = aggregator_context.aggregator_data.borrow_mut();
```

in every single native function